### PR TITLE
Set expected in-memory version when decoding unstructured objects from etcd

### LINKED ISSUE #72493: TestValidateOnlyStatus flakes with `apiVersion: Invalid value`
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1135464872302088192
```
--- FAIL: TestValidateOnlyStatus (1.61s)
I0102 19:03:43.523081  125200 establishing_controller.go:84] Shutting down EstablishingController
I0102 19:03:43.523219  125200 secure_serving.go:156] Stopped listening on 127.0.0.1:46393
I0102 19:03:43.523225  125200 naming_controller.go:295] Shutting down NamingConditionController
I0102 19:03:43.523093  125200 crd_finalizer.go:254] Shutting down CRDFinalizer
I0102 19:03:43.523081  125200 customresource_discovery_controller.go:214] Shutting down DiscoveryController
testserver.go:141: runtime-config=map[api/all:true]
testserver.go:142: Starting apiextensions-apiserver on port 36085...
testserver.go:160: Waiting for /healthz to be ok...
subresources_test.go:531: unexpected error: WishIHadChosenNoxu.mygroup.example.com "foo" is invalid: apiVersion: Invalid value: "mygroup.example.com/v1beta1": must be mygroup.example.com/v1
```

cc @mbohlool @sttts

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug
/kind flake

**What this PR does / why we need it**:
* Fixes in-memory versions and conversion for custom resources (and adds tests)
* Fixes defaulting of custom resource list items in list responses (and adds tests)

**Which issue(s) this PR fixes**:
Fixes #72493

**Does this PR introduce a user-facing change?**:
```release-note
Fixed a spurious error where update requests to the status subresource of multi-version custom resources would complain about an incorrect API version.
```

/cc @sttts @jpbetz @roycaihw 
/sig api-machinery

### CODE REVIEW

**Comment by liggitt** [comment]
successfully added a test that recreates the issue:

> TestWebhookConverterWithoutWatchCache/nontrivial-converter/check-3/getting_objects_created_as_v1alpha1:
>
> conversion_test.go:340: MultiVersion.stable.example.com "converted-v1alpha1" is invalid: apiVersion: Invalid value: "stable.example.com/v1beta1": must be stable.example.com/v1alpha1

**Comment by liggitt** [comment]
/priority critical-urgent

this fix is required for webhook conversion

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go [hunk 1]
```diff
@@ -177,12 +177,19 @@ type StatusREST struct {
...
 
 // Get retrieves the object from the storage. It is required to support Patch.
 func (r *StatusREST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
-	return r.store.Get(ctx, name, options)
+	o, err := r.store.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+	if u, ok := o.(*unstructured.Unstructured); ok {
+		shallowCopyObjectMeta(u)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go]: this is the important line?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go]: No, setting the kind in New(), and preserving it in the etcd3 store when instantisting the object we decode into. This was just a bug I found while here

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 2]
```diff
@@ -683,11 +686,14 @@ func (s *store) watch(ctx context.Context, key string, rv string, pred storage.S
 	return s.watcher.Watch(ctx, key, int64(rev), recursive, pred)
 }
 
-func (s *store) getState(getResp *clientv3.GetResponse, key string, v reflect.Value, ignoreNotFound bool) (*objState, error) {
+func (s *store) getState(getResp *clientv3.GetResponse, key string, v reflect.Value, hubGroupVersionKind schema.GroupVersionKind, ignoreNotFound bool) (*objState, error) {
 	state := &objState{
 		obj:  reflect.New(v.Type()).Interface().(runtime.Object),
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: this New is very odd.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 3]
```diff
@@ -683,11 +686,14 @@ func (s *store) watch(ctx context.Context, key string, rv string, pred storage.S
...
 		meta: &storage.ResponseMeta{},
 	}
+	// Set the hubGroupVersionKind into the new instance
+	// This is needed by unstructured decoders to know what version we want and convert appropriately
+	state.obj.GetObjectKind().SetGroupVersionKind(hubGroupVersionKind)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: this feels like bandaid. It works, but the real issue is that we create state.obj above via reflection. Instead this has to go through the strategy's New method.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: plumbing the New() method from the strategy required changing Storage interfaces, which I really don't want to do. Reworked this to make the `copy+set GVK` behavior part of the Unstructured interface and checked for that here

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: yes

**Comment by liggitt** [comment]
need to fix up the dummy data in the etcd create unit test… will repush shortly

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 4]
```diff
@@ -685,9 +685,15 @@ func (s *store) watch(ctx context.Context, key string, rv string, pred storage.S
 
 func (s *store) getState(getResp *clientv3.GetResponse, key string, v reflect.Value, ignoreNotFound bool) (*objState, error) {
 	state := &objState{
-		obj:  reflect.New(v.Type()).Interface().(runtime.Object),
 		meta: &storage.ResponseMeta{},
 	}
+
+	if u, ok := v.Addr().Interface().(runtime.Unstructured); ok {
```
Comment (dims) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: check using `CanAddr()`? test seems to be panic'ing

```
--- FAIL: TestCreateWithKeyExist (0.10s)
panic: reflect.Value.Addr of unaddressable value [recovered]
	panic: reflect.Value.Addr of unaddressable value
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 5]
```diff
@@ -783,7 +789,13 @@ func decode(codec runtime.Codec, versioner storage.Versioner, value []byte, objP
 
 // appendListItem decodes and appends the object (if it passes filter) to v, which must be a slice.
 func appendListItem(v reflect.Value, data []byte, rev uint64, pred storage.SelectionPredicate, codec runtime.Codec, versioner storage.Versioner) error {
-	obj, _, err := codec.Decode(data, nil, reflect.New(v.Type().Elem()).Interface().(runtime.Object))
+	var into runtime.Object
+	if u, ok := v.Addr().Interface().(runtime.Unstructured); ok {
```
Comment (dims) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Same as above

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: all real places that call this use `conversion.EnforcePtr(...)` which ensures addressability

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: all real places that call this use `conversion.EnforcePtr(...)` which ensures addressability

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: switched the test to use that as well

**Review by roycaihw — COMMENTED** [review]
/lgtm

the fix looks good to me. I have some followup questions

- do we want to cherrypick this for noop conversion and alpha webhook conversion?
- does it worth adding TypeMeta validation for old object in main resource as well?
- add test that exercises `UnstructuredList` conversion?

**Comment by liggitt** [comment]
> * do we want to cherrypick this for noop conversion and alpha webhook conversion?

I could see picking this back for no-op conversion. I wouldn't do it for an alpha feature.

> * does it worth adding TypeMeta validation for old object in main resource as well?

I don't think so. We aren't validating typemeta in the old object in the subresource, it just happened that the old object's typemeta was used to overwrite the incoming object's, and therefore passed through validation. If we ever got in a state where 

> * add test that exercises `UnstructuredList` conversion?

sure

**Comment by liggitt** [comment]
looks like defaulting has a similar issue. it flakes on master, and when I disable the watch cache in the integration test, it fails 100% of the time. will run that test with and without the watch cache, and dig into fixing that in this PR as well

/hold

**Comment by liggitt** [comment]
> looks like defaulting has a similar issue when I disable the watch cache in the integration test, it fails 100% of the time. will run that test with and without the watch cache, and dig into fixing that in this PR as well

There were a couple issues here:

**Issue the 1st**
The `TestCustomResourceDefaulting` integration test was not correct. It did the following:
* set up a single-version custom resource
* set up defaults for `spec.c` and `status.c` fields
* made a status update without specifying `status.c` and verified a get of the resulting object contained both `spec.c` and `status.c` set
* removed the defaults
* asserted the `spec.c` field should disappear from `get` responses

However, since defaulting is applied to the object in the incoming request *and* the object read from etcd, the status update combining the `status` stanza of the incoming object (with `status.c` defaulted) and the `spec` stanza of the existing object (with `spec.c` defaulted) *should* have persisted both fields in etcd. Even when the defaults were removed from the schema, those persisted values should remain.

How then, was this test passing in master? Soldier on, dear reader.

**Issue the 2nd**
Defaults were not being applied to custom resources in list responses. This was due to an incorrect short-circuit in `versioning.codec#Decode`, which skipped defaulting (incorrectly) along with conversion (correctly) when the provided object to decode into was used as-is:

https://github.com/kubernetes/kubernetes/blob/ca12f110470c6c649202f5229daeb044143f9451/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L116-L138

Only conversion should have been skipped there (there's no need to convert an object to itself). The result was that custom resource list responses did not have defaults applied to items.

A notable consumer of list responses is a lister/watcher that feeds the watch cache. That caused items in the watch cache from the initial list to be missing default values. Because updates attempt to use the last observed item from the watch cache as the "old object", that meant that *usually* the old object did not have defaults, so the `spec.c` field of the old object was *not* defaulted, and the incorrect test assertion that `spec.c` would disappear from get requests once the default was removed proved to be what actually happened, because `spec.c` never got persisted.

**The fix**
The fix correctly applies defaults on decode, and custom resource defaulting tests were added for the following operations:
* with the watch cache enabled:
  * get
  * get cached (resourceVersion=0)
  * list
  * list cached (resourceVersion=0)
* with the watch cache disabled:
  * get
  * get cached (resourceVersion=0)
  * list
  * list cached (resourceVersion=0)
  * watch

Additionally, a multi-version CRD was used, so that we can set different defaults for the storage version and REST API version, to make sure both phases are getting exercised.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go [hunk 6]
```diff
@@ -179,32 +275,116 @@ func TestCustomResourceDefaulting(t *testing.T) {
...
+	mustNotExist(foo.Object, [][]string{{"spec", "c"}, {"status", "c"}})
+
+	t.Logf("Updating status, expecting 'c' to be set in status only")
+	if foo, err = fooClient.UpdateStatus(foo, metav1.UpdateOptions{}); err != nil {
+		t.Fatal(err)
+	}
+	mustExist(foo.Object, [][]string{{"spec", "a"}, {"spec", "b"}, {"status", "a"}, {"status", "b"}, {"status", "c"}})
+	mustNotExist(foo.Object, [][]string{{"spec", "c"}})
+
+	t.Logf("Removing 'a', 'b' and `c` properties from the REST version. Expecting that 'c' goes away in spec, but not in status. 'a' and 'b' were peristed.")
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go]: persisted

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go [hunk 7]
```diff
@@ -179,32 +275,116 @@ func TestCustomResourceDefaulting(t *testing.T) {
...
+			)
+		case <-time.After(wait.ForeverTestTimeout):
+			t.Fatal("timed out without getting watch event")
+		}
+	}
+
+	t.Logf("Add 'c' default to the REST version, remove it from the storage version, and wait until GET no longer sees it in both status and spec")
+	addDefault("v1beta1", "c", "C")
+	removeDefault("v1beta2", "c")
+	if err := wait.PollImmediate(100*time.Millisecond, 20*time.Second, func() (bool, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go]: nit: wait.InfiniteTestTimeout

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go [hunk 8]
```diff
@@ -133,6 +126,14 @@ func (c *codec) Decode(data []byte, defaultGVK *schema.GroupVersionKind, into ru
...
 		}
 
+		// Short-circuit conversion if the into object is same object
+		if into == obj {
+			if isVersioned {
+				return versioned, gvk, nil
+			}
+			return into, gvk, nil
+		}
+
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go]: lgtm

File: pkg/printers/internalversion/printers_test.go [hunk 9]
```diff
@@ -235,6 +235,8 @@ func testPrinter(t *testing.T, printer printers.ResourcePrinter, unmarshalFunc f
 		TypeMeta:   metav1.TypeMeta{APIVersion: "v1", Kind: "Pod"},
 		ObjectMeta: metav1.ObjectMeta{Name: "foo"},
 	}
+	// our decoder defaults, so we should default our expected object as well
+	legacyscheme.Scheme.Default(obj)
```
Comment (sttts) [inline pkg/printers/internalversion/printers_test.go]: independent?

Comment (liggitt) [reply pkg/printers/internalversion/printers_test.go]: fixing the bug in versioning.codec that was skipping defaulting made this test fail… the expected object was missing defaults that were present in the decoded object

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go [hunk 10]
```diff
@@ -179,31 +275,115 @@ func TestCustomResourceDefaulting(t *testing.T) {
...
 	mustExist(foo.Object, [][]string{{"spec", "a"}, {"spec", "b"}, {"status", "a"}, {"status", "b"}})
 
-	t.Logf("Add 'c' default and wait until GET sees it in both status and spec")
-	addDefault("c", "C")
-	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+	t.Logf("Add 'c' default to the storage version and wait until GET sees it in both status and spec")
+	addDefault("v1beta2", "c", "C")
+
+	t.Logf("wait until GET sees 'c' in both status and spec")
+	if err := wait.PollImmediate(100*time.Millisecond, 20*time.Second, func() (bool, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go]: these 20 second I meant.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go [hunk 11]
```diff
@@ -179,31 +275,115 @@ func TestCustomResourceDefaulting(t *testing.T) {
...
+		return true, nil
 	}); err != nil {
 		t.Fatal(err)
 	}
 	mustExist(foo.Object, [][]string{{"spec", "a"}, {"spec", "b"}, {"spec", "c"}, {"status", "a"}, {"status", "b"}, {"status", "c"}})
 
-	t.Logf("Updating status, expecting 'c' to be set in spec and status")
-	if foo, err = fooClient.UpdateStatus(foo, metav1.UpdateOptions{}); err != nil {
+	t.Logf("wait until GET sees 'c' in both status and spec of cached get")
+	if err := wait.PollImmediate(100*time.Millisecond, 20*time.Second, func() (bool, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go]: and these

**Pull request merged**